File: go/indicators/donmak/modifiedexponentialmovingaverage/modifiedexponentialmovingaverage.go
package modifiedexponentialmovingaverage

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
)

// computeVelocityCoefficients computes FIR coefficients for the first derivative
// of a degree-degree polynomial fit evaluated at the most recent point (Lagrange
// basis, order=1).
func computeVelocityCoefficients(degree int) []float64 {
	nPoints := degree + 1
	coefficients := make([]float64, 0, nPoints)

	for i := 0; i < nPoints; i++ {
		denom := 1.0

		for j := 0; j < nPoints; j++ {
			if j != i {
				denom *= float64(j - i)
			}
		}

		others := make([]int, 0, degree)

		for j := 0; j < nPoints; j++ {
			if j != i {
				others = append(others, j)
			}
		}

		numerator := 0.0

		for ell := 0; ell < len(others); ell++ {
			term := 1.0

			for m := 0; m < len(others); m++ {
				if m != ell {
					term *= float64(others[m])
				}
			}

			numerator += term
		}

		coefficients = append(coefficients, numerator/denom)
	}

	return coefficients
}

// ModifiedExponentialMovingAverage is Don Mak's Modified Exponential Moving Average (MEMA / MEMA-D).
//
// It is a reduced-lag EMA that adds the EMA's own polynomial velocity back to its
// output, compensating for smoothing delay:
//
//	MEMA(n) = EMA(n) + PFD(EMA, degree, order=1, stride=skip)
//
// Reference:
//
// Mak, Don K. (2006). Mathematical Techniques in Financial Market Trading. Ch 4.2.
type ModifiedExponentialMovingAverage struct {
	mu sync.RWMutex

	core.LineIndicator

	degree  int
	skip    int
	nPoints int

	coefficients []float64

	emaAlpha       float64
	emaValue       float64
	emaInitialized bool

	buf      []float64
	bufSize  int
	bufPos   int
	bufCount int

	primed bool
}

// NewModifiedExponentialMovingAverage returns an instance of the indicator created using supplied parameters.
//
//nolint:funlen
func NewModifiedExponentialMovingAverage(p *Params) (*ModifiedExponentialMovingAverage, error) {
	const (
		invalid        = "invalid modified exponential moving average parameters"
		fmts           = "%s: %s"
		fmtw           = "%s: %w"
		defaultPeriod  = 6
		defaultDegree  = 3
		defaultSkip    = 1
	)

	period := p.Period
	if period == 0 {
		period = defaultPeriod
	}

	degree := p.Degree
	if degree == 0 {
		degree = defaultDegree
	}

	skip := p.Skip
	if skip == 0 {
		skip = defaultSkip
	}

	if period < 2 {
		return nil, fmt.Errorf(fmts, invalid, "period should be >= 2")
	}

	if degree < 2 {
		return nil, fmt.Errorf(fmts, invalid, "degree should be >= 2")
	}

	if skip < 1 {
		return nil, fmt.Errorf(fmts, invalid, "skip should be >= 1")
	}

	bc := p.BarComponent
	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	qc := p.QuoteComponent
	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	tc := p.TradeComponent
	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	var (
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	mnemonic := fmt.Sprintf("mema(%d,%d,%d%s)", period, degree, skip, core.ComponentTripleMnemonic(bc, qc, tc))
	desc := "Modified exponential moving average " + mnemonic

	bufSize := degree*skip + 1

	mema := &ModifiedExponentialMovingAverage{
		degree:       degree,
		skip:         skip,
		nPoints:      degree + 1,
		coefficients: computeVelocityCoefficients(degree),
		emaAlpha:     2.0 / (float64(period) + 1.0),
		buf:          make([]float64, bufSize),
		bufSize:      bufSize,
	}

	mema.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, mema.Update)

	return mema, nil
}

// IsPrimed indicates whether the indicator is primed.
func (s *ModifiedExponentialMovingAverage) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes the output data of the indicator.
func (s *ModifiedExponentialMovingAverage) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.ModifiedExponentialMovingAverage,
		s.LineIndicator.Mnemonic,
		s.LineIndicator.Description,
		[]core.OutputText{
			{Mnemonic: s.LineIndicator.Mnemonic, Description: s.LineIndicator.Description},
		},
	)
}

// Update updates the indicator given the next sample value and returns the filter output.
func (s *ModifiedExponentialMovingAverage) Update(sample float64) float64 {
	s.mu.Lock()
	defer s.mu.Unlock()

	// EMA recursion (seed at first sample).
	if !s.emaInitialized {
		s.emaValue = sample
		s.emaInitialized = true
	} else {
		s.emaValue = s.emaAlpha*sample + (1.0-s.emaAlpha)*s.emaValue
	}

	// Store EMA value in the ring buffer.
	s.buf[s.bufPos] = s.emaValue
	s.bufPos = (s.bufPos + 1) % s.bufSize
	s.bufCount++

	if s.bufCount < s.bufSize {
		s.primed = false

		return math.NaN()
	}

	s.primed = true

	// Read EMA values at stride positions and compute the velocity correction.
	velocity := 0.0
	for k := 0; k < s.nPoints; k++ {
		offset := k * s.skip
		idx := ((s.bufPos-1-offset)%s.bufSize + s.bufSize) % s.bufSize
		velocity += s.coefficients[k] * s.buf[idx]
	}

	return s.emaValue + velocity
}

File: ts/indicators/don-mak/modified-exponential-moving-average/modified-exponential-moving-average.ts
import { buildMetadata } from '../../core/build-metadata';
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { LineIndicator } from '../../core/line-indicator';
import { ModifiedExponentialMovingAverageParams } from './params';

/**
 * Computes FIR coefficients for the first derivative of a degree-`degree`
 * polynomial fit evaluated at the most recent point (Lagrange basis, order=1).
 */
function computeVelocityCoefficients(degree: number): number[] {
  const nPoints = degree + 1;
  const coefficients: number[] = [];

  for (let i = 0; i < nPoints; i++) {
    let denom = 1;
    for (let j = 0; j < nPoints; j++) {
      if (j !== i) {
        denom *= j - i;
      }
    }

    const others: number[] = [];
    for (let j = 0; j < nPoints; j++) {
      if (j !== i) {
        others.push(j);
      }
    }

    let numerator = 0;
    for (let ell = 0; ell < others.length; ell++) {
      let term = 1;
      for (let m = 0; m < others.length; m++) {
        if (m !== ell) {
          term *= others[m];
        }
      }
      numerator += term;
    }

    coefficients.push(numerator / denom);
  }

  return coefficients;
}

/** Function to calculate the mnemonic of a __ModifiedExponentialMovingAverage__ indicator. */
export const modifiedExponentialMovingAverageMnemonic = (params: ModifiedExponentialMovingAverageParams): string => {
  const period = Math.floor(params.period ?? 6);
  const degree = Math.floor(params.degree ?? 3);
  const skip = Math.floor(params.skip ?? 1);
  const cm = componentTripleMnemonic(params.barComponent, params.quoteComponent, params.tradeComponent);

  return `mema(${period},${degree},${skip}${cm})`;
};

/**
 * ModifiedExponentialMovingAverage is Don Mak's Modified Exponential Moving Average (MEMA / MEMA-D).
 *
 * It is a reduced-lag EMA that adds the EMA's own polynomial velocity back to its
 * output, compensating for smoothing delay:
 *
 *   MEMA(n) = EMA(n) + PFD(EMA, degree, order=1, stride=skip)
 *
 * Reference:
 *
 * Mak, Don K. (2006). Mathematical Techniques in Financial Market Trading. Ch 4.2.
 */
export class ModifiedExponentialMovingAverage extends LineIndicator {

  private readonly degree: number;
  private readonly skip: number;
  private readonly nPoints: number;
  private readonly coefficients: number[];

  private readonly emaAlpha: number;
  private emaValue = 0;
  private emaInitialized = false;

  private readonly buf: number[];
  private readonly bufSize: number;
  private bufPos = 0;
  private bufCount = 0;

  public constructor(params: ModifiedExponentialMovingAverageParams) {
    super();

    const period = Math.floor(params.period ?? 6);
    const degree = Math.floor(params.degree ?? 3);
    const skip = Math.floor(params.skip ?? 1);

    if (period < 2) {
      throw new Error('period should be >= 2');
    }
    if (degree < 2) {
      throw new Error('degree should be >= 2');
    }
    if (skip < 1) {
      throw new Error('skip should be >= 1');
    }

    this.mnemonic = modifiedExponentialMovingAverageMnemonic(params);
    this.description = 'Modified exponential moving average ' + this.mnemonic;
    this.barComponent = params.barComponent;
    this.quoteComponent = params.quoteComponent;
    this.tradeComponent = params.tradeComponent;

    this.degree = degree;
    this.skip = skip;
    this.nPoints = degree + 1;
    this.coefficients = computeVelocityCoefficients(degree);

    this.emaAlpha = 2 / (period + 1);

    this.bufSize = degree * skip + 1;
    this.buf = new Array<number>(this.bufSize).fill(0);
    this.primed = false;
  }

  public metadata(): IndicatorMetadata {
    return buildMetadata(
      IndicatorIdentifier.ModifiedExponentialMovingAverage,
      this.mnemonic,
      this.description,
      [{ mnemonic: this.mnemonic, description: this.description }],
    );
  }

  public update(sample: number): number {
    // EMA recursion (seed at first sample).
    if (!this.emaInitialized) {
      this.emaValue = sample;
      this.emaInitialized = true;
    } else {
      this.emaValue = this.emaAlpha * sample + (1 - this.emaAlpha) * this.emaValue;
    }

    // Store EMA value in the ring buffer.
    this.buf[this.bufPos] = this.emaValue;
    this.bufPos = (this.bufPos + 1) % this.bufSize;
    this.bufCount++;

    if (this.bufCount < this.bufSize) {
      this.primed = false;
      return Number.NaN;
    }

    this.primed = true;

    // Read EMA values at stride positions and compute the velocity correction.
    let velocity = 0;
    for (let k = 0; k < this.nPoints; k++) {
      const offset = k * this.skip;
      const idx = ((this.bufPos - 1 - offset) % this.bufSize + this.bufSize) % this.bufSize;
      velocity += this.coefficients[k] * this.buf[idx];
    }

    return this.emaValue + velocity;
  }
}
